### ISSUE TITLE
Separate up/down stairs

### ISSUE THREAD

**Issue by MichaelMackus** [LambdaHack/LambdaHack#231]
From what I can tell, the out of the box way to generate stairs is via a "staircase" which is just a place with up & down stairs. Is there no way to randomly generate separated up/down stairs? If not, we should look into adding this since most roguelikes are navigated using this method.

**Comment by Mikolaj** [comment]
Agreed. Content creator can already add stairs at will, by changing `TileKind` and `PlaceKind`. However I suspect problems if he does.

E.g., even every floor tile the content creator defines may have an embedded item with the effect `Ascend True` or `Ascend False` (BTW, it's worth checking if '>' key can then be redefined in Input content to activate the tile under feet and perhaps even only downstairs; I'm not sure about the latter, so may be worth a separate issue, slightly related to #152; I think this functionality may have been here in the past, but can probably be re-implemented better now). 

For a tame example, using already existing content, stairs can be added to the random filler tiles of any cave, with some probability, and then we can see what happens. A similar example is game exit added randomly to one of the cave edges (level "fences", the one using the tile group `AIRLOCK_FENCE`) here, so it's just a matter of adding a dozen of characters:

https://github.com/AllureOfTheStars/Allure/blob/236219a4f31dee4855f794d5793b922d1b3dec25/GameDefinition/Content/TileKind.hs#L1202

I fully expect an assertion failure somewhere, strange results (e.g., downstairs on deepest level would probably just teleport) and violation of the current invariants in level generation. One invariant is that stairs on adjacent levels are aligned (re position and re kind of stairs, e.g., if you enter a lift, you don't exit from a stair on another level). Others are `cminStairDist` (already only kept on the best effort basis) and `cmaxStairsNum` (see their haddocks). We'd need to tackle all three problems, but the last one, the moral problem of violated invariants can be breached even without experiments. 

At this stage of the engine development, e.g., having your planned roguelike in mind, do we want to relax the invariants? How much? E.g., stairs generated via staircase places still uphold the invariants, but random stairs don't? Add an extra flag in `Cave Kind` that declares how relaxed stair generation can be and assertions elsewhere enforce it and perhaps some stages of dungeon generation are guided by to provide best effects given the expected random tile (containing stairs) placement later on?

IIRC, tile spacial alignment is determined by position on the `lstair` lists. I guess adding stairs randomly does not change the lists. Perhaps it should and that could be the first lead into the codebase.

**Comment by MichaelMackus** [comment]
I did try adding single stairs but it failed with an assertion pretty much
instantly. In regards to short term fixes, I agree that adding a flag to
remove the constraints would work fine for my game. The only concern I have
is that I would want to enforce at least 1 down stair and probably 1 up
stair - adding them to random tiles would mean that this enforcement
wouldn't always be met (unless I make stair tiles fairly common).

**Comment by MichaelMackus** [comment]
> A related question: do we plan non-persistent levels any time soon?

I was actually originally planning on having non-persistent levels, or at least up/down stairs that don't *always* lead back to the previous level. I'm not sure how possible this is in the current engine, or even if it might make implementation of up/down stairs easier, so I didn't really mention it since it isn't a high priority for me ATM.

**Comment by Mikolaj** [comment]
Most important of all this easily answers the question: where and to what extent exactly we honour the spatially aligned stairs invariant. With that out of the way, it may be just a matter of fixing some assertions to ignore checks if `persistent==False` or similarly easy tweaks to code assumptions. 

The price, however, is refactoring level generation so that it can be performed at any time. But it would improve the quality of code and enable effects (spells) that completely rebuild a level (or many) also in the persistent case. A non-persistent branch in a persistent dungeon is a nice variation, too, as is the other way around.

Yet another answer to "what about space alignment invariant", the most difficult, but the most useful, may be to have another effect, different to `Ascend` that acts as an inter-level teleporter. Perhaps it should have an extra flag that tells if the destination should always be by a (matching vs. lift/stair/shaft/etc.) stairs in the opposite direction or should be a completely random walkable non-lava position. It's easy to implement --- basically remove a lot of stuff from the code for `Ascend`. One important thing to decide is if normal stairs should lead to such a teleporter (hard) and if it should lead to normal stars or teleporters or both (each easy).

BTW, I've just re-read your initial comment. Do you mean that in LH a staircase needs to have stairs to the upper level *and* stairs to the lower level? That's not true, though that's often the case to help the player quickly navigate the persistent layers and to save space for more interesting rooms.

**Comment by Mikolaj** [comment]
Hmm, strange, I did this in LH:

```
 stairsUp = TileKind
   { tsymbol  = '<'
   , tname    = "staircase up"
-  , tfreq    = [(STAIRCASE_UP, 9), (ORDINARY_STAIRCASE_UP, 1)]
+  , tfreq    = [(STAIRCASE_UP, 9), (ORDINARY_STAIRCASE_UP, 1), (EMPTY_SET_LIT, 800)]
```

and it works in crawl scenario without problems. How did you get the assertion failure?

Agreed that depending on RNG to generate at least 1 stairs is silly. I think many roguelikes does it via rooms (places) even just to ensure some ceremony, becuase stairs are important so underscoring them or making them harder to find or safer to arrive at makes sense.. E.g., Angband does, I think. I'm not sure how early it determines where to put that place, though. I guess either early, or restart level generation if it doesn't fit (which is silly IMHO, too). What's wrong with using places?

**Comment by MichaelMackus** [comment]
From my own experience, I generate up & down stairs based solely on RNG positions on the map. I might have a distance constraint, but I do not do so based on the room. However, my roguelikes are usually very basic and do not rely much on the grid/BSP method beyond the basic dungeon-layout. Constraining this to room-type could also have further problems if we expand the engine to support other dungeon/cave generation beyond grid-based maps.

**Comment by MichaelMackus** [comment]
Also, I feel like the aligning of up & down stairs should at least be optional if that is what is causing a lot of these issues. It is not essential for most games IMO, but is definitely a nice-to-have.

**Comment by Mikolaj** [comment]
Agreed. I think, however you add the stairs tile, it should work fine if you add it to the `lstair` field of `Level`. Again. what crashed for  you when you did it via random filler tiles features?
